src/main/kotlin/Stopwatch.kt

```diff
@@ -1,37 +1,42 @@
-import java.util.concurrent.CopyOnWriteArraySet
+import io.reactivex.Observable
+import io.reactivex.Scheduler
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.PublishSubject
+import io.reactivex.subjects.Subject
+import java.util.concurrent.TimeUnit
 import java.util.concurrent.atomic.AtomicLong
 
-interface Stopwatch {
-
-    interface StopwatchListener {
-        fun onValueChange(value: Long)
-    }
 
-    fun resumeStopwatch(stopwatchListener: StopwatchListener)
-    fun pauseStopwatch(stopwatchListener: StopwatchListener)
-
-    class Impl(private val ticker: Ticker) : Stopwatch, Ticker.TickListener {
-
-        private val listeners = CopyOnWriteArraySet<StopwatchListener>()
-        private val value = AtomicLong(0)
+interface Stopwatch {
 
-        override fun onTick() {
-            val newValue = value.incrementAndGet()
-            listeners.forEach { it.onValueChange(newValue) }
+    val value: Subject<Long>
+    fun pause()
+    fun resume()
+
+    class Impl(private val scheduler: Scheduler) : Stopwatch {
+
+        override val value =  BehaviorSubject.create<Long>().toSerialized()
+        private val consumersCountChange = PublishSubject.create<Long>().toSerialized()
+        private val actualValueCounter = AtomicLong()
```

Comment:
I think you don’t need that. It is fine if clients count drops to zero and the time starts from the start.

Reply:
I could remove that, but then it would create a weird situation, where, if functions are called in this order: "resume - resume - pause", the stopwatch would not stop after pause, but it will reset (because it's changed from 2 (after second resume) to 1 (after pause) and thus not filtered out by .filter, and so it is switched from one Observable.interval to another Observable.interval (=reset when it doesn't seems logical for it to reset)).
Is that okay?